src/rai_bench/rai_bench/tool_calling_agent/tasks/custom_interfaces.py

```diff
@@ -158,328 +199,865 @@ def available_tools(self) -> List[BaseTool]:
-        target_y: float = 2.0,
-        target_z: float = 3.0,
-        initial_gripper_state: bool = True,
-        final_gripper_state: bool = False,
-        frame_id: str = "base_link",
+        validators: Optional[List[Validator]] = None,
+        logger: logging.Logger | None = None,
     ) -> None:
-        super().__init__(
-            service, service_args, validators=validators, task_args=task_args
-        )
-        self.expected_initial_gripper_state = initial_gripper_state
-        self.expected_final_gripper_state = final_gripper_state
-        self.expected_target_pose = PoseStamped(
-            header=Header(frame_id=frame_id),
-            pose=Pose(
-                position=Point(x=target_x, y=target_y, z=target_z),
-                orientation=Quaternion(x=0.0, y=0.0, z=0.0, w=1.0),
-            ),
-        )
+        self.target_x = target_x
+        self.target_y = target_y
+        self.target_z = target_z
+        self.initial_gripper_state = initial_gripper_state
+        self.final_gripper_state = final_gripper_state
+        if validators is None:
+            # Default validator for this task
+            get_manipulator_interface_subtask = CheckArgsToolCallSubTask(
+                expected_tool_name="get_ros2_message_interface",
+                expected_args={"msg_type": MANIPULATOR_SERVICE_TYPE},
+            )
+            call_manipulator_service_subtask = CheckServiceFieldsToolCallSubTask(
+                expected_tool_name="call_ros2_service",
+                expected_service=MANIPULATOR_SERVICE,
+                expected_service_type=MANIPULATOR_SERVICE_TYPE,
+                expected_fields={
+                    "target_pose.pose.position.x": target_x,
+                    "target_pose.pose.position.y": target_y,
+                    "target_pose.pose.position.z": target_z,
+                    "initial_gripper_state": initial_gripper_state,
+                    "final_gripper_state": final_gripper_state,
+                },
+            )
+            validators = [
+                OrderedCallsValidator(
+                    subtasks=[
+                        get_manipulator_interface_subtask,
+                        call_manipulator_service_subtask,
+                    ]
+                )
+            ]
+        super().__init__(validators, task_args, logger)
 
     def get_base_prompt(self) -> str:
-        pos = self.expected_target_pose.pose.position
         return (
-            f"Call service '{self.service}' to move manipulator to pose "
-            f"({pos.x}, {pos.y}, {pos.z}) with initial_gripper={self.expected_initial_gripper_state}, "
-            f"final_gripper={self.expected_final_gripper_state}."
+            f"Call service '{MANIPULATOR_SERVICE}' to move manipulator to pose "
+            f"({self.target_x}, {self.target_y}, {self.target_z}) with initial gripper state {self.initial_gripper_state} "
+            f"and final gripper state {self.final_gripper_state}."
```

Comment:
Would it make sense to add more indirect prompt? 
For example: "grap an object from (x,y,z)", or "please give me an object 'object_name'. It's located at (x,y,z)"
I think validators will be the same for such case.

Reply:
A direct prompt makes sense as well, as a simple test / baseline but the usecase would be more of a ros2 cli assistant. But we should test more embodied AI usecases as well.